src/DCON/Ode.jl

```diff
@@ -0,0 +1,722 @@
+#         # Big slice: u = ua[:, mpert+1:2*mpert,:]
+#         for i = 1:mpert, j = 1:mpert, k = 1:2
+#             u[i, j, k] = ua[i, mpert+j, k]
+#         end
+#     end
+#     u_save .= u
+#     psi_save = psifac
+#     msol = mpert
+#     neq = 4 * mpert * msol
+
+#     # Compute conditions at next singular surface
+#     while true
+#         ising += 1
+#         if ising > msing || psilim < sing[ising ≤ msing ? ising : msing].psifac
+#             break
+#         end
+#         odet.q = sing[ising].q
+#         if mlow ≤ nn*q && mhigh ≥ nn*q
+#             break
+#         end
+#     end
+# This needs to be fixed up
+#     if ising > msing || psilim < sing[ising ≤ msing ? ising : msing].psifac
+#         m1 = round(Int, ctrl.nn*intr.qlim) + round(Int, sign(one, ctrl.nn*equil.sq.fs1[mpsi, 4]))
+#         psimax = psilim * (1-eps)
+#         next_ = "finish"
+#     else
+#         m1 = round(Int, nn*sing[ising].q)
+#         psimax = sing[ising].psifac - singfac_min/abs(nn*sing[ising].q1)
+#         next_ = "cross"
+#     end
+
+#     # Set up integrator parameters
+#     istep = 0
+#     istate = 1
+#     iopt = 1
+#     itask = 5
+#     itol = 2
+#     mf = 10
+#     istate = 1
+
+#     # Set up work arrays
+#     liw = length(iwork)    # iwork must exist!
+#     lrw = 22 + 64*mpert*msol
+#     rwork = zeros(r8, lrw)
+#     atol  = zeros(r8, mpert, msol, 2)
+#     fixfac = zeros(r8, msol, msol)
+#     iwork = zeros(Int, liw)  # or readjust as needed
+#     rwork[1] = psimax
+#     rwork[5] = dpsi
+#     rwork[11] = rwork[5]
+
+#     # Terminate, or in Julia just return (no need for RETURN)
+#     return nothing  # or could return a struct with all these values, for a more Julian approach
+# end
+
+function ode_ideal_cross(ising::Int, odet::OdeState, equil::Equilibrium.PlasmaEquilibrium, intr::DconInternal, ctrl::DconControl)
+    # ...existing code...
+
+    # Fixup solution at singular surface
+    if ctrl.verbose
+        println("psi = $(intr.sing[ising].psifac), q = $(intr.sing[ising].q)")
+    end
+    ode_unorm(true)    
+
+    # Write asymptotic coefficients before reinit
+    if ctrl.bin_euler
+   #     sing_get_ca(ising, odet.psifac, u, ca)
+   #     write(euler_bin_unit, 4)
+   #     write(euler_bin_unit, intr.sing[ising].psifac, intr.sing[ising].q, intr.sing[ising].q1)
+   #     write(euler_bin_unit, msol)
+   #     write(euler_bin_unit, ca)
+    end
+
+    # Re-initialize
+    psi_old = odet.psifac
+    ipert0 = round(Int, ctrl.nn * intr.sing[ising].q) - intr.mlow + 1
+    dpsi = intr.sing[ising].psifac - odet.psifac
+    odet.psifac = intr.sing[ising].psifac + dpsi
+    sing_get_ua(ising, odet.psifac, ua)
+    if !con_flag
+        u[:, index[1], :] .= 0  # originally u(ipert0,:,:) = 0
+    end
+    sing_der(neq, psi_old, u, du1)
+    sing_der(neq, odet.psifac, u, du2)
+    u .= u .+ (du1 .+ du2) .* dpsi
+    if !con_flag
+        u[ipert0, :, :] .= 0
+        u[:, index[1], :] .= ua[:, ipert0 + mpert, :]
```

Comment:
Whenever you're slicing on the right-hand side, you allocate a new array. You want to use [`@view` or `@views`](https://docs.julialang.org/en/v1/base/arrays/#Views-(SubArrays-and-other-view-types))